Userland/Libraries/LibGfx/QOIWriter.cpp

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (c) 2022, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "QOIWriter.h"
+#include "QOICommon.h"
+#include <AK/ByteBuffer.h>
+#include <AK/Math.h>
+#include <AK/MemoryStream.h>
+#include <AK/TypedTransfer.h>
+#include <LibGfx/Bitmap.h>
+
+namespace Gfx {
+
+template<typename Stream = ByteBuffer>
+class QOIWriterState : public QOIState {
+public:
+    ErrorOr<void> start(u32 width, u32 height, bool has_alpha_channel)
+    {
+        if (m_pixels_to_encode != INVALID_PIXELS_TO_ENCODE)
+            return Error::from_string_literal("Trying to start an already initialized QOI writer"sv);
+
+        m_pixels_to_encode = static_cast<u64>(width) * static_cast<u64>(height);
+        VERIFY(m_pixels_to_encode != INVALID_PIXELS_TO_ENCODE);
+
+#if 0
+        // 5 is the worse case scenario where we need to encode QOI_OP_RGBA for all pixels
+        const u64 pixels_to_encode_size = m_pixels_to_encode * 5;
+        TRY(m_stream.try_reserve(sizeof(QOIHeader) + pixels_to_encode_size + sizeof(QOI_END_MARKER)));
+#endif
+
+        TRY(try_encode_qoi_header(width, height, has_alpha_channel));
+        return {};
+    }
+
+    ErrorOr<void> encode(Color const pixel)
+    {
+        if (m_pixels_to_encode == INVALID_PIXELS_TO_ENCODE)
+            return Error::from_string_literal("Trying to encode using a not initialized QOI writer"sv);
+        if (m_pixels_to_encode == 0)
+            return Error::from_string_literal("Trying to encode using a QOI writer that should be finished instead"sv);
+        --m_pixels_to_encode;
+
+        const i8 dr = pixel.red() - get_previous_pixel().red();
+        const i8 dg = pixel.green() - get_previous_pixel().green();
+        const i8 db = pixel.blue() - get_previous_pixel().blue();
+        const i8 da = pixel.alpha() - get_previous_pixel().alpha();
+
+        false
```

Comment:
Please no 😁

Reply:
No to `db` `da`? I used the names from the specs, so one can refer to it more easily.
About the `false`, I agree it is a controversial style, but it helped me to quickly disable and trouble shoot parts of the encoder. Now That I have a working base, I can iterate on it.

Reply:
No to this cute short-circuit trick.